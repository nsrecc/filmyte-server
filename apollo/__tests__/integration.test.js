import { createTestClient } from 'apollo-server-testing';
import { createTestServer } from 'apollo/__tests__/utils';
import { mockConfigurationResponse } from 'apollo/datasources/__tests__/tmdbMocks/configurationMock';
import {
  mockMovieGenreListResponse,
  mockTvGenreListResponse,
} from 'apollo/datasources/__tests__/tmdbMocks/genreListMock';
import { mockMovieDetailsResponse } from 'apollo/datasources/__tests__/tmdbMocks/movieDetailsMock';
import { mockTvDetailsResponse } from 'apollo/datasources/__tests__/tmdbMocks/tvDetailsMock';
import {
  GET_CONFIGURATION,
  GET_GENRES_BY_MEDIA_TYPE,
  GET_MOVIE_DETAILS,
  GET_TV_DETAILS,
} from 'apollo/__tests__/graphql/tmdb';

/**
 * --- Apollo Server Integration Tests ---
 *
 * Integration tests for testing the APIs as a black box. The scope of integration tests is to
 * receive a graphql query, and mock-fetch the data for the query using the correct datasource API.
 *
 * More information: https://www.apollographql.com/docs/apollo-server/testing/testing/
 */
describe('apollo server integration tests', () => {
  // after each test, clear the calls and instances of all mocks
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Query GET_CONFIGURATION', () => {
    it('should query configuration and return expected result', async () => {
      // create instance of test server which uses existing typeDefs, resolvers, and datasource APIs
      const { server, tmdbAPI } = createTestServer();
      // mock tmdbAPI get function with mock response and give it a name for reference
      tmdbAPI.get = jest.fn().mockName('tmdbAPI.get');
      tmdbAPI.get.mockReturnValue(mockConfigurationResponse);

      // use the test server to create a query function
      const { query } = createTestClient(server);
      // run query against the server and snapshot the output
      const res = await query({ query: GET_CONFIGURATION });

      expect(tmdbAPI.get).toHaveBeenCalledTimes(1);
      expect(res).toMatchSnapshot();
    });
  });

  describe('Query GET_GENRES_BY_MEDIA_TYPE', () => {
    it('should query genres by media type and return expected result', async () => {
      const { server, tmdbAPI } = createTestServer();
      tmdbAPI.get = jest.fn().mockName('tmdbAPI.get');
      tmdbAPI.get.mockReturnValueOnce(mockMovieGenreListResponse);
      tmdbAPI.get.mockReturnValueOnce(mockTvGenreListResponse);

      const { query } = createTestClient(server);
      const res = await query({ query: GET_GENRES_BY_MEDIA_TYPE });

      expect(tmdbAPI.get).toHaveBeenCalledTimes(2);
      expect(res).toMatchSnapshot();
    });
  });

  describe('Query GET_MOVIE_DETAILS', () => {
    it('should query movie details by id and return expected result', async () => {
      const { server, tmdbAPI } = createTestServer();
      tmdbAPI.get = jest.fn().mockName('tmdbAPI.get');
      tmdbAPI.get.mockReturnValue(mockMovieDetailsResponse);
      const { id: movieId } = mockMovieDetailsResponse;

      const { query } = createTestClient(server);
      const res = await query({ query: GET_MOVIE_DETAILS, variables: { movieId } });

      expect(tmdbAPI.get).toHaveBeenCalledTimes(1);
      expect(res).toMatchSnapshot();
    });
  });

  describe('Query GET_TV_DETAILS', () => {
    it('should query tv details by id and return expected result', async () => {
      const { server, tmdbAPI } = createTestServer();
      tmdbAPI.get = jest.fn().mockName('tmdbAPI.get');
      tmdbAPI.get.mockReturnValue(mockTvDetailsResponse);
      const { id: tvId } = mockTvDetailsResponse;

      const { query } = createTestClient(server);
      const res = await query({ query: GET_TV_DETAILS, variables: { tvId } });

      expect(tmdbAPI.get).toHaveBeenCalledTimes(1);
      expect(res).toMatchSnapshot();
    });
  });
});
