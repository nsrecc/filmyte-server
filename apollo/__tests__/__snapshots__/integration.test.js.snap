// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apollo server integration tests Query GET_CONFIGURATION should query configuration and return expected result 1`] = `
Object {
  "data": Object {
    "configuration": Object {
      "change_keys": Array [
        "adult",
        "air_date",
        "also_known_as",
        "alternative_titles",
        "biography",
        "birthday",
        "budget",
        "cast",
        "certifications",
        "character_names",
        "created_by",
        "crew",
        "deathday",
        "episode",
        "episode_number",
        "episode_run_time",
        "freebase_id",
        "freebase_mid",
        "general",
        "genres",
        "guest_stars",
        "homepage",
        "images",
        "imdb_id",
        "languages",
        "name",
        "network",
        "origin_country",
        "original_name",
        "original_title",
        "overview",
        "parts",
        "place_of_birth",
        "plot_keywords",
        "production_code",
        "production_companies",
        "production_countries",
        "releases",
        "revenue",
        "runtime",
        "season",
        "season_number",
        "season_regular",
        "spoken_languages",
        "status",
        "tagline",
        "title",
        "translations",
        "tvdb_id",
        "tvrage_id",
        "type",
        "video",
        "videos",
      ],
      "images": Object {
        "backdrop_sizes": Array [
          "w300",
          "w780",
          "w1280",
          "original",
        ],
        "base_url": "http://image.tmdb.org/t/p/",
        "logo_sizes": Array [
          "w45",
          "w92",
          "w154",
          "w185",
          "w300",
          "w500",
          "original",
        ],
        "poster_sizes": Array [
          "w92",
          "w154",
          "w185",
          "w342",
          "w500",
          "w780",
          "original",
        ],
        "profile_sizes": Array [
          "w45",
          "w185",
          "h632",
          "original",
        ],
        "secure_base_url": "https://image.tmdb.org/t/p/",
        "still_sizes": Array [
          "w92",
          "w185",
          "w300",
          "original",
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`apollo server integration tests Query GET_GENRES_BY_MEDIA_TYPE should query genres by media type and return expected result 1`] = `
Object {
  "data": Object {
    "genresByMediaType": Object {
      "movieGenres": Array [
        Object {
          "id": 28,
          "name": "Action",
        },
        Object {
          "id": 12,
          "name": "Adventure",
        },
        Object {
          "id": 16,
          "name": "Animation",
        },
        Object {
          "id": 35,
          "name": "Comedy",
        },
        Object {
          "id": 80,
          "name": "Crime",
        },
        Object {
          "id": 99,
          "name": "Documentary",
        },
        Object {
          "id": 18,
          "name": "Drama",
        },
        Object {
          "id": 10751,
          "name": "Family",
        },
        Object {
          "id": 14,
          "name": "Fantasy",
        },
        Object {
          "id": 36,
          "name": "History",
        },
        Object {
          "id": 27,
          "name": "Horror",
        },
        Object {
          "id": 10402,
          "name": "Music",
        },
        Object {
          "id": 9648,
          "name": "Mystery",
        },
        Object {
          "id": 10749,
          "name": "Romance",
        },
        Object {
          "id": 878,
          "name": "Science Fiction",
        },
        Object {
          "id": 10770,
          "name": "TV Movie",
        },
        Object {
          "id": 53,
          "name": "Thriller",
        },
        Object {
          "id": 10752,
          "name": "War",
        },
        Object {
          "id": 37,
          "name": "Western",
        },
      ],
      "tvGenres": Array [
        Object {
          "id": 10759,
          "name": "Action & Adventure",
        },
        Object {
          "id": 16,
          "name": "Animation",
        },
        Object {
          "id": 35,
          "name": "Comedy",
        },
        Object {
          "id": 80,
          "name": "Crime",
        },
        Object {
          "id": 99,
          "name": "Documentary",
        },
        Object {
          "id": 18,
          "name": "Drama",
        },
        Object {
          "id": 10751,
          "name": "Family",
        },
        Object {
          "id": 10762,
          "name": "Kids",
        },
        Object {
          "id": 9648,
          "name": "Mystery",
        },
        Object {
          "id": 10763,
          "name": "News",
        },
        Object {
          "id": 10764,
          "name": "Reality",
        },
        Object {
          "id": 10765,
          "name": "Sci-Fi & Fantasy",
        },
        Object {
          "id": 10766,
          "name": "Soap",
        },
        Object {
          "id": 10767,
          "name": "Talk",
        },
        Object {
          "id": 10768,
          "name": "War & Politics",
        },
        Object {
          "id": 37,
          "name": "Western",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`apollo server integration tests Query GET_MOVIE_DETAILS should query configuration and return expected result 1`] = `
Object {
  "data": Object {
    "movieDetails": Object {
      "adult": false,
      "backdrop_path": "/xJHokMbljvjADYdit5fK5VQsXEG.jpg",
      "belongs_to_collection": null,
      "budget": 165000000,
      "genres": Array [
        Object {
          "id": 12,
          "name": "Adventure",
        },
        Object {
          "id": 18,
          "name": "Drama",
        },
        Object {
          "id": 878,
          "name": "Science Fiction",
        },
      ],
      "homepage": "http://www.interstellarmovie.net/",
      "id": 157336,
      "imdb_id": "tt0816692",
      "original_language": "en",
      "original_title": "Interstellar",
      "overview": "The adventures of a group of explorers who make use of a newly discovered wormhole to surpass the limitations on human space travel and conquer the vast distances involved in an interstellar voyage.",
      "popularity": 92.332,
      "poster_path": "/gEU2QniE6E77NI6lCU6MxlNBvIx.jpg",
      "production_companies": Array [
        Object {
          "id": 923,
          "logo_path": "/5UQsZrfbfG2dYJbx8DxfoTr2Bvu.png",
          "name": "Legendary Pictures",
          "origin_country": "US",
        },
        Object {
          "id": 9996,
          "logo_path": "/3tvBqYsBhxWeHlu62SIJ1el93O7.png",
          "name": "Syncopy",
          "origin_country": "GB",
        },
        Object {
          "id": 13769,
          "logo_path": null,
          "name": "Lynda Obst Productions",
          "origin_country": "",
        },
      ],
      "production_countries": Array [
        Object {
          "iso_3166_1": "GB",
          "name": "United Kingdom",
        },
        Object {
          "iso_3166_1": "US",
          "name": "United States of America",
        },
      ],
      "release_date": "2014-11-05",
      "revenue": 675120017,
      "runtime": 169,
      "spoken_languages": Array [
        Object {
          "iso_639_1": "en",
          "name": "English",
        },
      ],
      "status": "Released",
      "tagline": "Mankind was born on Earth. It was never meant to die here.",
      "title": "Interstellar",
      "video": false,
      "vote_average": 8.3,
      "vote_count": 24079,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
